# Text Processing, Tokenization, and N-gram Language Modeling

Implementations of email validation, minimum-edit-distance correction, rule-based/BPE/WordPiece tokenization, and smoothed N-gram language models on Persian TinyStories.

**Course:** Natural Language Processing, University of Tehran


# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال اول - <span dir="ltr">Regular Expression & Min Distance</span> (20)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال شما با نمونه‌هایی عملی از استفاده‌ی Regex و همینطور Minimum Distance مواجه می‌شوید. در بخش اول این سوال، شما باید ایمیل‌های قابل‌قبول را تشخیص داده و در بخش دوم سوال شما با استفاده از Minimum Distance یک سیستم Auto-Correction ساده را پیاده‌سازی خواهید کرد.


## <div style="text-align: center; direction: rtl; font-family: Vazir;">تشخیص ایمیل‌های قابل قبول با Regex</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
فایل emails.txt که در اختیار شما قرار داده‌شده، شامل تعدادی اسم به‌همراه ایمیل ثبت‌شده‌‌شان در یک سامانه می‌باشد.
<br>
از شما خواسته‌شده است تا ایمیل‌هایی که معتبر هستند را با استفاده از regex مشخص کنید.
<br>
ایمیل از دو بخش تشکیل می‌شود. که با @ از هم جدا می‌شوند. بخش اول (قبل از @) local-part نام دارد و بخش دوم domain.
<br>
منظور از ایمیل معتبر این است که موارد زیر در آن‌ها رعایت شده‌باشند:
<br>
۱. دو بخش ایمیل با یک و تنها یک @ از هم جدا شده‌باشند.
<br>
۲. در local-part هم نام و هم نام خانوادگی شخص وجود داشته باشد.
<br>
۳. در local-part تنها حروف انگلیسی، اعداد و کاراکترهای -،_ و . مجاز هستند. همچنین دو نقطه نمی‌توانند پشت هم بیایند.
<br>
۴. در بخش domain یک میزبان داریم و یک پسوند. میزبان و پسوند همیشه با یک نقطه از یکدیگر جدا می‌شوند. (میزبان می‌تواند در خود نقطه داشته باشد، اما دو نقطه‌ی متوالی در domain مجاز نیست.)
<br>
۵. پسوند از حروف انگلیسی تشکیل می‌شود و حداقل دو کاراکتر دارد.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- لیست ایمیل‌های قابل قبول موجود در فایل emails.txt
</p>

In [9]:
import re

def get_all_mail_records(mail_file_path='./emails.txt', item_seperator=', ', key_val_seperator='='):
    records = []
    with open(mail_file_path, 'r') as file:
        for line in file:
            records.append(line.strip())
    return records

# regex pattern for filtering
pattern = re.compile(
r'(?i)^name=(?P<first>[a-z]+)\s+(?P<last>[a-z]+),'
r'\s*email='
r'(?P<email>'
r'(?=[a-z0-9._-]*?(?P=first))'
r'(?=[a-z0-9._-]*?(?P=last))'
r'(?!.*\.\.)'
r'[a-z0-9](?:[a-z0-9._-]*[a-z0-9])?'
r'@'
r'[a-z0-9]+(?:\.[a-z0-9]+)*\.[a-z]{2,})$'
)

def get_filtered_emails(records, pattern=pattern):
    filtered_records = [record for record in records if pattern.match(record)]
    filtered_emails = [record.split('email=')[1] for record in filtered_records]
    return filtered_emails

records = get_all_mail_records()
filtered_mails = get_filtered_emails(records)
print('Filtered Emails:', *[f"{i}. {email}" for i, email in enumerate(filtered_mails, start=1)], sep='\n')

Filtered Emails:
1. milad.fattahi@uni.ir
2. parsa.golkar@host.ir
3. moradi.reza@uni.ir
4. niloufarheidari@lab.ir
5. mahsa_akbari@domain.com
6. ali.rezaei@example.com
7. zarei-farhad@host.ir
8. hamed.jalali@net.com
9. ramin.shafiei@project.net
10. sadeghi.laleh@work.ir
11. nouri.pouya@mail.ir
12. elhamdavari@office.net
13. behnam.khatibi@mail.org
14. sara.soleimani@host.ir
15. shirin_hashemi@domain.com
16. arman.khalili@school.ir
17. rahaesfandiari@company.com
18. kiana.ghasemi@uni.edu
19. navid_khodadadi@research.org
20. mehrdad.ebrahimi@example.org
21. raminshafiei@project.net
22. hamedjalali@domain.org


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در این بخش باید با استفاده از Regex یک تابع بنویسید که اسم یک شخص را به‌عنوان ورودی دریافت کند و درصورت وجود ایمیل معتبر برای این اسم در بین ایمیل‌های ثبت‌شده، ایمیل را برگرداند. در غیر این‌صورت، یک پیغام عدم وجود چاپ کند.
<br>
توجه: ممکن است ایمیل این اشخاص، تحت نام شخص دیگری ثبت شده باشد. کار شما این است که ایمیل معتبر این افراد را از بین تمام ایمیل‌ها پیدا کنید.
<br>
تابع را با ورودی‌های زیر اجرا کنید:
<span dir="ltr" style="text-align: left; display: block;">
name1 = Behnam Khatibi
</span>
<span dir="ltr" style="text-align: left; display: block;">
name2 = Mehrdad Ebrahimi
</span>
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- ایمیل‌های معتبر مربوط به افراد ذکر شده در توضیحات
</p>

In [10]:
def find_valid_mail(name, records):
    mails = [record.split('email=')[1] for record in records]
    first, last = name.split(' ')
    
    # regex for filtering
    pattern = re.compile(
    rf'(?i)(?P<email>('
    rf'(?=[a-z0-9._-]*?{first})'           
    rf'(?=[a-z0-9._-]*?{last})'            
    rf'(?!.*\.\.)'                         
    rf'[a-z0-9](?:[a-z0-9._-]*[a-z0-9])?'  
    rf'@'
    rf'[a-z0-9]+(?:\.[a-z0-9]+)*\.[a-z]{{2,}}))$'
    )
    
    valid_mails = [mail for mail in mails if pattern.match(mail)]
    if len(valid_mails):
        print('Valid Mails:', *valid_mails)
    else:
        print('No Valid Mails Found')

find_valid_mail('Behnam Khatibi', records)
find_valid_mail('Mehrdad Ebrahimi', records)
    

Valid Mails: behnam.khatibi@mail.org
Valid Mails: mehrdad.ebrahimi@example.org


## <div style="text-align: center; direction: rtl; font-family: Vazir;">پیاده‌سازی Auto-Correction با Minimum Edit Distance</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
۱. ابتدا الگوریتم levenshtein_distance را پیاده‌سازی کنید و با استفاده از آن minimum_distance بین کلمات زیر را به‌دست آورید.
<span dir="ltr" style="text-align: left; display: block;">
pair1 = "Athletic", "Atlantic"
</span>
<span dir="ltr" style="text-align: left; display: block;">
pair2 = "London", "Boston"
</span>
<span dir="ltr" style="text-align: left; display: block;">
pair3 = "Action", "Compact"
</span>
<span dir="ltr" style="text-align: left; display: block;">
pair3 = "", "Sting"
</span>
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- مقدار minimum distance بین جفت کلمه‌های داده‌شده در بخش توضیحات
</p>

In [4]:
# (Assuming cost of substitution is 2)
def minimum_distance(a, b, report=False):
    m, n = len(a), len(b)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    # dp initialization
    for i in range(m + 1):
        dp[i][0] = i
    for j in range(n + 1):
        dp[0][j] = j

    # dep update
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 2 # substitute
            dp[i][j] = min(
                dp[i - 1][j] + 1,      # delete
                dp[i][j - 1] + 1,      # insert
                dp[i - 1][j - 1] + cost
            )

    if report:
        print(f'Minimum Edit Distance between \'{a}\' and \'{b}\':', dp[m][n])
    return dp[m][n]


minimum_distance('Athletic', 'Atlantic', report=True)
minimum_distance('London', 'Boston', report=True)
minimum_distance('Action', 'Compact', report=True)
minimum_distance('', 'Sting', report=True)

Minimum Edit Distance between 'Athletic' and 'Atlantic': 4
Minimum Edit Distance between 'London' and 'Boston': 6
Minimum Edit Distance between 'Action' and 'Compact': 9
Minimum Edit Distance between '' and 'Sting': 5


5

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
۲. بعد از پیاده‌سازی minimum distance حال باید با استفاده از آن، جمله‌ی زیر را اصلاح املایی کنید. در این جمله تعدادی کلمه وجود دارند که املایشان نادرست است. یک لیست از املای صحیح کلمات که کلمات این جمله را نیز شامل می‌شوند در فایل vocab.txt موجود هستند.
<span dir="ltr" style="text-align: left; display: block;">
sentence_to_be_corrected = 
"helo studnts at the universty are wrting ther frst edit distnce algorthm in pythn, and they reely enjy it!"
</span>
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- اصلاح‌شده‌ی جمله‌ی داده‌شده در بخش توضیحات با کمک فایل vocab.txt
</p>

In [5]:
import string
STOPWORDS = ['a', 'an', 'the', 'and', 'or', 'but', 'if', 'in', 'on', 'at', 'to', 'for', 'of', 'by', 'with', 'from', 'as', 'is', 'it', 'this', 'that']

def get_vocab(path='./vocab.txt'):
    with open(path, 'r') as f:
        words_str = (f.read())
        return [x.strip() for x in words_str.split(',')]        

def get_correct_word(word, vocab):
    min_distances = [minimum_distance(word, vocab[i]) for i in range(len(vocab))]
    arg_min = min_distances.index(min(min_distances))
    return vocab[arg_min]

def correct_sentence(sentence, vocab):
    words = re.findall(r'\w+|[^\w\s]', sentence)
    for i in range(len(words)):
        if words[i] in list(string.punctuation) or words[i] in STOPWORDS:
            continue
        words[i] = get_correct_word(words[i], vocab)
    corrected_sentence = ' '.join(words)
    cleaned_corrected_sentence = re.sub(r'\s+([^\w\s])', r'\1', corrected_sentence)
    return cleaned_corrected_sentence


vocab = get_vocab()
sentence = "helo studnts at the universty are wrting ther frst edit distnce algorthm in pythn, and they reely enjy it!"
print(correct_sentence(sentence, vocab))

hello students at the university are writing their first edit distance algorithm in Python, and they really enjoy it!


# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال دوم - <span dir="ltr">Tokenization</span> (25)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال، با انواع مختلف توکنایزر و روش  پیاده‌سازی آن‌ها آشنا می‌شوید.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">Rule-based Tokenizer</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
با کمک دستورات regex یک توکنایزر بنویسید که متن را با استفاده از علائم نگارشی تقسیم کند.
<br>
سپس عملکرد آن را بر روی جملات داده شده زیر آزمایش کنید.
<span dir="ltr" style="text-align: left; display: block;">
"Hello, world! NLP is fun."
</span>
<span dir="ltr" style="text-align: left; display: block;">
"That U.S.A. poster-print costs $12.40..."
</span>
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- تعداد و لیست توکن‌های ایجادشده با توکنایزر صورت سوال برای هر جمله<br>
</p>

In [6]:
def tokenize_sentence(sentence):
    tokenization_pattern = r'\w+|[^\w\s]'
    tokens = re.findall(tokenization_pattern, sentence)
    return len(tokens), tokens

def report_tokenization_result(sentence):
    token_count, tokens = tokenize_sentence(sentence)
    print(f"Tokenization result for the sentence: \'{sentence}\':\nNumber of Tokens: {token_count}\nTokens: {tokens}\n")

report_tokenization_result("Hello, world! NLP is fun.")
report_tokenization_result("That U.S.A. poster-print costs $12.40...")


Tokenization result for the sentence: 'Hello, world! NLP is fun.':
Number of Tokens: 8
Tokens: ['Hello', ',', 'world', '!', 'NLP', 'is', 'fun', '.']

Tokenization result for the sentence: 'That U.S.A. poster-print costs $12.40...':
Number of Tokens: 18
Tokens: ['That', 'U', '.', 'S', '.', 'A', '.', 'poster', '-', 'print', 'costs', '$', '12', '.', '40', '.', '.', '.']



<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ایرادات این توکنایزر چیست؟
<br>
چند راه‌حل برای بهبود عملکرد این توکنایزر ارائه دهید.
<br>
یکی از آن‌ها را پیاده‌سازی کنید.
<br>
عملکرد توکنایزر جدید را بر روی جملات داده شده آزمایش کنید.
<span dir="ltr" style="text-align: left; display: block;">
"Hello, world! NLP is fun."
</span>
<span dir="ltr" style="text-align: left; display: block;">
"That U.S.A. poster-print costs $12.40..."
</span>
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- تعداد و لیست توکن‌های ایجادشده با توکنایزر بهبودیافته برای هر جمله
</p>

In [7]:
def tokenize_sentence_enhanced(sentence):
  tokenization_pattern = r"""
      \.\.\.                 # ...
    | [A-Za-z]\.[A-Za-z]\.(?:[A-Za-z]\.)*  # abbreviations
    | \$\d+(?:\.\d+)?        # for currency values like $12.40
    | \d+\.\d+               # for decimals 12.40
    | \w+(?:-\w+)*           # for normal words and words having '-' in them
    | [^\w\s]
  """
  tokens = re.findall(tokenization_pattern, sentence, re.VERBOSE)
  return len(tokens), tokens

def report_tokenization_result_enhaced(sentence):
    token_count, tokens = tokenize_sentence_enhanced(sentence)
    print(f"Tokenization result for the sentence: \'{sentence}\':\nNumber of Tokens: {token_count}\nTokens: {tokens}\n")

report_tokenization_result_enhaced("Hello, world! NLP is fun.")
report_tokenization_result_enhaced("That U.S.A. poster-print costs $12.40...")

Tokenization result for the sentence: 'Hello, world! NLP is fun.':
Number of Tokens: 8
Tokens: ['Hello', ',', 'world', '!', 'NLP', 'is', 'fun', '.']

Tokenization result for the sentence: 'That U.S.A. poster-print costs $12.40...':
Number of Tokens: 6
Tokens: ['That', 'U.S.A.', 'poster-print', 'costs', '$12.40', '...']



<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش اول:</b><br>
<br>
جواب:
<br>
حاصل توکنایز کردن جملۀ اول قابل قبول می‌باشد در حالی که برای جملۀ دوم همان‌طور که در خروجی نیز دیده می‌شود، ایراداتی از قبیل موارد زیر وجود دارد:
<br>
1- کلمۀ U.S.A به صورت شش توکن مجزا توکنایز شده است که بهتر است به صورت یک کلمۀ واحد باشد.
<br>
2- کلمۀ poster-print به صورت سه توکن مجزا توکنایز شده است که باز بهتر است کل کلمه به صورت یک کلمۀ واحد توکنایز شود.
<br>
3- کاراکتر ... به صورت سه توکن مجزای نقطه توکنایز شده است که بهتر است کل این کاراکتر به صورت یک توکن واحد درآید.
<br>
4- و...
<br>
یک راه حل می‌تواند به این صورت باشد که پشتیبانی از کارکترهایی مانند سه نقطه، کلمات مخفف که شامل نقاطی در میان کلمه می‌باشند و یا کماتی که در شامل یک کاراکتر '-' در میان خود می‌باشند را به الگوی regex اضافه کنیم تا عملیات توکنایز کردن برای این گروه از کلمات به صورت بهینه صورت گیرد. همچنین در صورتی که پشتیبانی از اعداد اعشاری به طور عمده‌تر، مقادیر پولی را اضافه کنیم تا کل عدد به صورت یک توکن درآید نیز می‌تواند کمک‌کننده باشد که این مورد به نحوۀ پیاده‌سازی منطق کدهای وابسته به این توکن‌ها نیز بستگی دارد.
<br>
پیاده‌سازی این موارد در سلول بالا قابل مشاهده است.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">BPE Tokenizer</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابتدا دیتاست TinyStories-Farsi را از HuggingFace لود کنید.
<a href="https://huggingface.co/datasets/taesiri/TinyStories-Farsi" target="_blank" rel="noopener noreferrer" style="color: #9EEAD2; text-decoration: underline;">
    (لینک دیتاست)
</a>
<br>
سپس از داده آموزش (train) آن جملات فارسی را در یک لیست اضافه کنید.
<br>
اکنون با استفاده از این مجموعه داده، یک توکنایزر BPE آموزش دهید. استفاده از کتابخانه‌های آماده مانعی ندارد.
<br>
عملکرد توکنایزر را روی جمله زیر آزمایش کنید:
<br>
"روزی یک مرد ثروتمند، پسر بچه کوچکش را بـه ده برد تا بـه او نشان دهد مردمی که در آنجا زندگی می‌کنند، چقدر فقیر هستند."
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- مجموعه داده آموزش<br>
- تعداد و لیست توکن‌های ایجادشده برای جمله
</p>

In [ ]:
import pandas as pd

splits = {'train': 'InProgress-TinyStoriesV2-GPT4-train-Translated-To-Farsi-w-Claude-2.0.csv', 'validation': 'TinyStoriesV2-GPT4-valid-Translated-To-Farsi-w-Claude-2.0.csv'}
df = pd.read_csv("hf://datasets/taesiri/TinyStories-Farsi/" + splits["train"])
print(df.columns)
df.to_parquet('./Data/Tiny_Stories_Farsi.parquet')

In [ ]:
import pandas as pd

cur_df = pd.read_parquet('./Data/Tiny_Stories_Farsi.parquet')
cur_df.drop(columns=['English'], inplace=True, axis=1)
print(cur_df.shape, cur_df.columns)
print(cur_df.iloc[0, :])

# create sentences
sentences = list(cur_df['Persian'])
print(sentences[0])

(124411, 1) Index(['Persian'], dtype='object')
Persian     یک‌روز یک پسربچه کوچولو به اسم بن بود. بن دوس...
Name: 0, dtype: object
 یک‌روز یک پسربچه کوچولو به اسم بن بود. بن دوست داشت دنیای اطرافش را کشف کند. او چیزهای شگفت‌انگیز زیادی دید، مثل وازه‌های زیبایی که در یک مغازه به نمایش گذاشته شده بودند. یک روز هنگامی‌که بن از مغازه رد می‌شد یک وازه بسیار ویژه را دید. وقتی بن آن را دید مبهوت شد! 
او گفت:«واو، این واقعا یک وازه شگفت‌انگیز است! آیا می‌توانم آن را بخرم؟»
فروشنده لبخند زد و گفت: «بله، البته می‌توانی. می‌توانی آن را به خانه ببری و به همه دوستانت نشان دهی که چقدر شگفت‌انگیز است!»
پس بن وازه را به خانه برد و از آن بسیار مفتخر بود! او دوستانش را صدا کرد و وازه شگفت‌انگیز را به آنها نشان داد. همه دوستانش فکر می‌کردند که وازه زیباست و نمی‌توانستند باور کنند که بن چقدر خوش‌شانس است.
و این‌گونه بود که بن یک وازه شگفت‌انگیز را در مغازه پیدا کرد!


In [12]:
from tokenizers import Tokenizer, models, trainers, pre_tokenizers, normalizers
from tokenizers.normalizers import Sequence, NFD, Lowercase, StripAccents, Replace
from tokenizers.pre_tokenizers import BertPreTokenizer
import re

# Model configuration
tokenizer = Tokenizer(models.BPE())

tokenizer.normalizer = Sequence([
    NFD(),  
    Replace("‌", " "),  
    Replace("\u200c", " "), 
    StripAccents(), 
])
tokenizer.pre_tokenizer = BertPreTokenizer()

trainer = trainers.BpeTrainer(
    vocab_size=32000,
    min_frequency=2,
    show_progress=True,
    special_tokens=["<unk>", "<s>", "</s>", "<pad>", "<mask>"]
)

In [ ]:
# Train tokenizer
tokenizer.train_from_iterator(sentences, trainer)
tokenizer.save("./Tokenizer/tiny_stories_farsi_bpe.json")

In [14]:
# Usage
encoded = tokenizer.encode(".روزی یک مرد ثروتمند، پسر بچه کوچکش را به ده برد تا به او نشان دهد مردمی که در آنجا زندگی می‌کنند، چقدر فقیر هستند")
print(f'Number of tokens: {len(encoded.tokens)}')
print(f'Tokens: {encoded.tokens}')

Number of tokens: 29
Tokens: ['.', 'روزی', 'یک', 'مرد', 'ثروتمند', '،', 'پسر', 'بچه', 'کوچکش', 'را', 'به', 'ده', 'برد', 'تا', 'به', 'او', 'نشان', 'دهد', 'مردمی', 'که', 'در', 'انجا', 'زندگی', 'می', 'کنند', '،', 'چقدر', 'فقیر', 'هستند']


## <div style="text-align: center; direction: rtl; font-family: Vazir;">Wordpiece Tokenizer</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
درباره Wordpiece Tokenizer تحقیق کنید.
<br>
نحوه آموزش این توکنایزر را به طور دقیق شرح دهید و سپس آن را با BPE مقایسه کنید.
<br>
اکنون یک توکنایزر Wordpiece بر روی مجموعه داده خود آموزش دهید. استفاده از کتابخانه‌های آماده مانعی ندارد.
<br>
عملکرد توکنایزر را روی جمله زیر آزمایش کنید:
<br>
"روزی یک مرد ثروتمند، پسر بچه کوچکش را بـه ده برد تا بـه او نشان دهد مردمی که در آنجا زندگی می‌کنند، چقدر فقیر هستند."
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- تعداد و لیست توکن‌های ایجادشده برای جمله
</p>

In [15]:
# Model configuration
tokenizer = Tokenizer(models.WordPiece(unk_token="<unk>"))

tokenizer.normalizer = Sequence([
    NFD(), 
    Replace("‌", " "),  
    Replace("\u200c", " "),  
    StripAccents(), 
])
tokenizer.pre_tokenizer = BertPreTokenizer()

trainer = trainers.WordPieceTrainer(
    vocab_size=32000,
    min_frequency=2,
    show_progress=True,
    special_tokens=["<unk>", "<s>", "</s>", "<pad>", "<mask>"]
)

In [ ]:
# Train tokenizer
tokenizer.train_from_iterator(sentences, trainer)
tokenizer.save("./Tokenizer/tiny_stories_farsi_wordpiece.json")

In [17]:
encoded = tokenizer.encode(".روزی یک مرد ثروتمند، پسر بچه کوچکش را به ده برد تا به او نشان دهد مردمی که در آنجا زندگی می‌کنند، چقدر فقیر هستند")
print(f'Number of tokens: {len(encoded.tokens)}')
print(f'Tokens: {encoded.tokens}')

Number of tokens: 29
Tokens: ['.', 'روزی', 'یک', 'مرد', 'ثروتمند', '،', 'پسر', 'بچه', 'کوچکش', 'را', 'به', 'ده', 'برد', 'تا', 'به', 'او', 'نشان', 'دهد', 'مردمی', 'که', 'در', 'انجا', 'زندگی', 'می', 'کنند', '،', 'چقدر', 'فقیر', 'هستند']


<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش سوم:</b><br>
<br>
عملکرد آموزش توکناریز wordpiece بر روی یک مجموعه داده شامل مراحل زیر است:
<br>
1- ابتدا این توکنایزر، برای هر کاراکتر موجود در مجموعۀ داده یک entry مجزا در لیست لغات در نظر می‌گیرد.
<br>
2- سپس، برای هر کلمۀ موجود در دیتای فراهم شده، حالات شکاندن کلمات به عناصر vocabulary تا آن لحظه را در نظر می‌گیرد و سپس آن ادغامی را انجام می‌دهد که سبب می‌شود احتمال کل ظهور کلمات موجود در دیتا، با استفاده از روش unigram بیشینه شود. پس از تبیین مرج مورد نظر برای این هدف، لیست vocabulary به روز رسانی می‌شود.
به بیان بهتر، در این روش هدف این است که واژگان به گونه‌ای انتخاب شوند که احتمال وقوع متن داده بیشینه شود که این احتمال در اصل، به صورت لگاریتم رابطه رو به رو محاسبه می‌شود: L(V) = ∏ P(w | V) که در آن: P(w | V) = ∏ P(t_i) و w = t_1, t_2, ... می‌باشد.
<br>
3- پس از به روز رسانی لیست لغات، این مراحل تا رسیدن لیست لغات به حد مشخص شده تکرار می‌شوند.
<br>
<br>
در روش BPE، مرحلۀ آماده‌سازی اولیه لغات همانند روش wordpiece است. تفاوت این روش با روش wordpiece در این است که در این روش، برای انجام مرج لغات، آن مرجی انجام می‌شود که بیشترین نرخ تکرار در دادۀ آموزش را دارد. پس از اینکه پرتکرارترین مرج شناخته شد و ادغام صورت پذیرفت و لیست لغات به روز رسانی شد، مجددا همانند روش wordpiece، این مراحل تا رسیدن لیست لغات به حد تعیین شده ادامه می‌یابد.
<br>
<br>
برای فهم بهتر، مثال زیر را در نظر می‌گیریم که قصد داریم یک توکنایزر بر روی لیست دادۀ ["low", "lower", "lowest"] آموزش بدهیم. بدین منظور، در هر دو روش ابتدا لیست لغات را به صورت کاراکترهای موجود در دادۀ آموزش در می‌آوریم: { l, o, w, e, r, s, t }
<br>
سپس، باید ادغام مورد نظر را انجام دهیم. در روش BPE، ابتدا تعداد جفت‌های مجاور از لیست کلمات در دادۀ آموزش محاسبه می‌شود که در مثال ما به صورت زیر است:
<br>
3: ('l','o')
<br>
3: ('o','w')
<br>
2: ('w','e')
<br>
1: ('e','r')
<br>
1: ('e','s')
<br>
1: ('s','t')
<br>
حال با توجه به داشتن بیشترین تکرار، دو لغت l و o در توکنایزر مرج می‌شوند و لیست لغات به روز رسانی می‌شود و این مراحل تا رسیدن به حد تعیین شده برای لیست لغات ادامه می‌یابد.
<br>
<br>
در روش wordpiece، برای انجام ادغام، احتمال بروز کلمات برای پس از انجام هر ادغام اندازه‌گیری می‌شود و آن ادغامی انجام می‌گیرد که این احتمال را بیشینه کند. در این مقال، به صورت احتمالی ابتدا l + o -> lo و سپس lo + w -> low می‌باشد و این مراحل تا رسیدن به حد تعیین شده ادامه می‌یابند.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">Tokenization Visualization</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
با استفاده از ابزار 
<a href="https://tiktokenizer.vercel.app/" target="_blank" rel="noopener noreferrer" style="color: #9EEAD2; text-decoration: underline;">
    tiktokenizer
</a>
توکن‌های تولید شده برای جمله زیر را در هر یک از مدل‌های gpt2 و gpt4 و Meta-Llama-3-8B را مشاهده و تفاوت‌ها را گزارش کنید.
<br>
"روزی یک مرد ثروتمند، پسر بچه کوچکش را بـه ده برد تا بـه او نشان دهد مردمی که در آنجا زندگی می‌کنند، چقدر فقیر هستند."
<br>
سپس در مورد توکنایزر استفاده شده در هر یک تحقیق کنید. به نظر شما علت تفاوت نتیجه آن‌ها در چیست؟
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- توکن‌های ایجادشده با هر توکنایزر برای جمله
</p>

<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش چهارم:</b>
<br>
خروجی توکنایزرها به صورت زیر است:
<br>
<img src="./Assets/gpt2.png" alt="gpt4" width="40%" height="40%">
<img src="./Assets/gpt4.png" alt="gpt4" width="40%" height="40%">
<img src="./Assets/llama.png" alt="gpt4" width="40%" height="40%">
<br>
<br>
علت تفاوت در خروجی توکنایرزها مربوط به نوع توکنایزر هر کدام و دادۀ آموزش آن‌ها می‌باشد. در مدل gpt2 از توکنایزر BPE  استفاده می‌شود و زبان دادۀ آموزش آن اکثرا انگلیسی بوده و به همین دلیل بر روی توکنایز کردن جملات فارسی عملکرد خوبی ندارد و همان‌طور که دیده می‌شود، اکثر کلمات را به صورت چند توکن در نظر می‌گیرد.
<br>
در مدل gpt4 نیز از یک ورژن بهبود یافتۀ توکنایزر BPE استفاده شده که بر روی زبان‌های بیشتری نسبت به gpt2 آموزش دیده و در نتیجه کمی عملکرد بهتری در توکنایز کرددن جملات فارسی دارد، برای مثال در gpt2، 'در' به صورت  دو توکن در نظر گرفته شده در حالی که در gpt4 یک توکن واحد در نظر گرفته شده است و همین سبب کاهش تعداد توکن شده است.
<br>
در مدل لاما، از توکنایزر SentencePiece Unigram استفاده شده که به صورت گسترده بر روی زبان‌های مختلف آموزش دیده و به همین دلیل در توکنایز کردن جملات فارسی عملکرد خوبی دارد و همان‌طور که دیده می‌شود، توانسته است اکثر کلمات را به درستی توکنایز کند و همین مورد سبب کاهش چشمگیر تعداد توکن در این توکنایزر شده است.
</p>


# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال سوم - <span dir="ltr">N-gram Language Modeling</span> (55)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال، با N-gram Language Modeling و آن را پیاده‌سازی می‌کنید، با استفاده از آن به تولید متن می‌پردازید. سپس با معیار perplexity و نحوه کاربرد آن آشنا می‌شوید. در نهایت الگوریتم‌های smoothing و با کاربردهای آن آشنا می‌شوید. را پیاده‌سازی می‌کنید
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">Data cleaning & Tokenization</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
مجموعه داده 
<a href="https://huggingface.co/datasets/taesiri/TinyStories-Farsi" target="_blank" rel="noopener noreferrer" style="color: #9EEAD2; text-decoration: underline;">
    TinyStories-Farsi
</a>
- که در سوال اول نیز از آن استفاده کردید - را لود کنید.
<br>
ابتدا دادگان فارسی در مجموعه آموزش (train) آن را تمیز کرده و پیش‌پردازش‌های مورد نیاز را بر روی آن انجام دهید. (با بررسی داده‌ها، مشخص‌کنید که این دادگان به چه پیش‌پردازش‌هایی نیاز دارند.)
<br>
سپس با استفاده از BPE Tokenizer که بر روی این دادگان آموزش‌داده‌اید، دادگان پردازش‌شده را توکنایز کنید.
<br>
توکن‌های یک جمله را به انتخاب خود، چاپ کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b>
<br>
- مجموعه داده تمیزشده فارسی
<br>
- مجموعه داده توکنایز شده فارسی
<br>
- توکن‌های یک جمله‌ی دلخواه
</p>

In [ ]:
import numpy as np

train_df = pd.read_parquet('./Data/Tiny_Stories_Farsi.parquet')
train_df.drop(columns=['English'], inplace=True, axis=1)
print(train_df.shape, train_df.columns)

(124411, 1) Index(['Persian'], dtype='object')


In [21]:
rand_idx = np.random.randint(0, len(train_df))
print(train_df.iloc[rand_idx]['Persian'])

 یک روزی، در سرزمینی دور، معبدی روشن وجود داشت. معبد بسیار زیبا و پر نور بود. خیلی‌ها برای تماشای نور معبد را زیارت می‌کردند. 
یک روز، دختری کوچک به نام امی رفت معبد را ببیند. او مشکل بزرگی دید. نور نبود! معبد تاریک و بی‌درخشش شده بود. امی می‌خواست کمک کند و مشکل را حل کند.  
امی اطراف را نگاه کرد و جعبه بزرگی دید. فکر کرد نور درون جعبه است. جعبه را باز کرد، ولی خالی بود. بعد امی به آسمان نگاه کرد و نور را دید! خورشید بود! خورشید پشت ابر بزرگی پنهان شده بود. ابر کنار رفت و نور به معبد برگشت. معبد دوباره پر نور شد و همه خوشحال بودند.


In [24]:
from parsivar import Normalizer

def preprocess_text(text, normalizer):
    # Normalize text
    text = normalizer.normalize(text)
    # Replace half space with space for clarity
    text = text.replace('\u200c', ' ')
    # Remove Latin letters
    text = re.sub(r'[A-Za-z0-9]', '', text)
    # Remove stray symbols or quotes
    text = re.sub(r'[“”"\'`]', '', text)
    # Collapse multiple spaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

def preprocess_data(df, preprocess_function, normalizer, col_name='Persian'):
    cleaned_df = df.copy()
    cleaned_df[col_name] = cleaned_df[col_name].apply(lambda x: preprocess_function(x, normalizer))      
    return cleaned_df

normalizer = Normalizer(statistical_space_correction=True)
cleaned_df = preprocess_data(train_df, preprocess_text, normalizer)
print(cleaned_df.shape, cleaned_df.columns)
print(cleaned_df.iloc[rand_idx]['Persian'])

(124411, 1) Index(['Persian'], dtype='object')
یک روزی ، در سرزمینی دور ، معبدی روشن وجود داشت . معبد بسیار زیبا و پر نور بود . خیلی ها برای تماشای نور معبد را زیارت می کردند . یک روز ، دختری کوچک به نام امی رفت معبد را ببیند . او مشکل بزرگی دید . نور نبود ! معبد تاریک و بی درخشش شده بود . امی می خواست کمک کند و مشکل را حل کند . امی اطراف را نگاه کرد و جعبه بزرگی دید . فکر کرد نور درون جعبه است . جعبه را باز کرد ، ولی خالی بود . بعد امی به آسمان نگاه کرد و نور را دید ! خورشید بود ! خورشید پشت ابر بزرگی پنهان شده بود . ابر کنار رفت و نور به معبد برگشت . معبد دوباره پر نور شد و همه خوشحال بودند .


In [ ]:
bpe_tokenizer = Tokenizer.from_file('./Tokenizer/tiny_stories_farsi_bpe.json')
print(bpe_tokenizer)

Tokenizer(version="1.0", truncation=None, padding=None, added_tokens=[{"id":0, "content":"<unk>", "single_word":False, "lstrip":False, "rstrip":False, ...}, {"id":1, "content":"<s>", "single_word":False, "lstrip":False, "rstrip":False, ...}, {"id":2, "content":"</s>", "single_word":False, "lstrip":False, "rstrip":False, ...}, {"id":3, "content":"<pad>", "single_word":False, "lstrip":False, "rstrip":False, ...}, {"id":4, "content":"<mask>", "single_word":False, "lstrip":False, "rstrip":False, ...}], normalizer=Sequence(normalizers=[NFD(), Replace(pattern=String("‌"), content=" "), Replace(pattern=String("‌"), content=" "), StripAccents()]), pre_tokenizer=BertPreTokenizer(), post_processor=None, decoder=None, model=BPE(dropout=None, unk_token=None, continuing_subword_prefix=None, end_of_word_suffix=None, fuse_unk=False, byte_fallback=False, ignore_merges=False, vocab={"<unk>":0, "<s>":1, "</s>":2, "<pad>":3, "<mask>":4, ...}, merges=[("ا", "ن"), ("ن", "د"), ("ب", "ا"), ("م", "ی"), ("ر", 

In [26]:
def bpe_tokenize_data(cleaned_df, tokenizer, col_name='Persian'):
    tokenized_df = cleaned_df.copy()
    tokenized_df[col_name] = cleaned_df[col_name].apply(lambda x: tokenizer.encode(x).tokens)
    return tokenized_df

tokenized_df = bpe_tokenize_data(cleaned_df, bpe_tokenizer)
print(tokenized_df.shape, tokenized_df.columns)
print(type(tokenized_df.iloc[rand_idx]['Persian']))
print(tokenized_df.iloc[rand_idx]['Persian'])

(124411, 1) Index(['Persian'], dtype='object')
<class 'list'>
['یک', 'روزی', '،', 'در', 'سرزمینی', 'دور', '،', 'معبدی', 'روشن', 'وجود', 'داشت', '.', 'معبد', 'بسیار', 'زیبا', 'و', 'پر', 'نور', 'بود', '.', 'خیلی', 'ها', 'برای', 'تماشای', 'نور', 'معبد', 'را', 'زیارت', 'می', 'کردند', '.', 'یک', 'روز', '،', 'دختری', 'کوچک', 'به', 'نام', 'امی', 'رفت', 'معبد', 'را', 'ببیند', '.', 'او', 'مشکل', 'بزرگی', 'دید', '.', 'نور', 'نبود', '!', 'معبد', 'تاریک', 'و', 'بی', 'درخشش', 'شده', 'بود', '.', 'امی', 'می', 'خواست', 'کمک', 'کند', 'و', 'مشکل', 'را', 'حل', 'کند', '.', 'امی', 'اطراف', 'را', 'نگاه', 'کرد', 'و', 'جعبه', 'بزرگی', 'دید', '.', 'فکر', 'کرد', 'نور', 'درون', 'جعبه', 'است', '.', 'جعبه', 'را', 'باز', 'کرد', '،', 'ولی', 'خالی', 'بود', '.', 'بعد', 'امی', 'به', 'اسمان', 'نگاه', 'کرد', 'و', 'نور', 'را', 'دید', '!', 'خورشید', 'بود', '!', 'خورشید', 'پشت', 'ابر', 'بزرگی', 'پنهان', 'شده', 'بود', '.', 'ابر', 'کنار', 'رفت', 'و', 'نور', 'به', 'معبد', 'برگشت', '.', 'معبد', 'دوباره', 'پر', 'نور', 'شد', 'و',

In [35]:
print('Raw sentence:', train_df.iloc[rand_idx]['Persian'], sep='\n')
print('Tokenized:', tokenized_df.iloc[rand_idx]['Persian'], sep='\n')

Raw sentence:
 یک روزی، در سرزمینی دور، معبدی روشن وجود داشت. معبد بسیار زیبا و پر نور بود. خیلی‌ها برای تماشای نور معبد را زیارت می‌کردند. 
یک روز، دختری کوچک به نام امی رفت معبد را ببیند. او مشکل بزرگی دید. نور نبود! معبد تاریک و بی‌درخشش شده بود. امی می‌خواست کمک کند و مشکل را حل کند.  
امی اطراف را نگاه کرد و جعبه بزرگی دید. فکر کرد نور درون جعبه است. جعبه را باز کرد، ولی خالی بود. بعد امی به آسمان نگاه کرد و نور را دید! خورشید بود! خورشید پشت ابر بزرگی پنهان شده بود. ابر کنار رفت و نور به معبد برگشت. معبد دوباره پر نور شد و همه خوشحال بودند.
Tokenized:
['یک', 'روزی', '،', 'در', 'سرزمینی', 'دور', '،', 'معبدی', 'روشن', 'وجود', 'داشت', '.', 'معبد', 'بسیار', 'زیبا', 'و', 'پر', 'نور', 'بود', '.', 'خیلی', 'ها', 'برای', 'تماشای', 'نور', 'معبد', 'را', 'زیارت', 'می', 'کردند', '.', 'یک', 'روز', '،', 'دختری', 'کوچک', 'به', 'نام', 'امی', 'رفت', 'معبد', 'را', 'ببیند', '.', 'او', 'مشکل', 'بزرگی', 'دید', '.', 'نور', 'نبود', '!', 'معبد', 'تاریک', 'و', 'بی', 'درخشش', 'شده', 'بود', '.', 'امی', 'می'

In [30]:
# save for later use
tokenized_df.to_parquet('./Data/bpe_tokenized_tiny_stories_farsi.parquet')

<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش اول:</b><br>
از آن‌جایی که داده به طور کلی داده‌ای تمیز می‌باشد و همچنین با توجه به اینکه  باید به توکنایزر آموزش دیدۀ BPE تحویل داده شود، نیازی به پیش‌پردازش خاص و یا با لود بالایی نیست؛ بدین منظور برای انجام پیش‌پردازش، از normalizer کتابخانۀ parsivar استفاده شده که مواردی مانند حروف تکراری، فاصله‌های پشت سر هم، نیم‌فاصله و موارد مشابه را مدیریت می‌کند. همچنین، برای اطمینان بیشتر از مراحل پیش پردازش، تعدادی از این مراحل به صورت جداگانه هم در مراحل بعد اعمال شده اند که در تابع پیش پردازش قابل رویت است. بدین ترتیب، متن حاصل می‌تواند به عنوان ورودی توکنایزر BPE استفاده شود.
می‌توان تغییرات در اثر ‌پیش‌پردازش را با مقایسۀ خروجی سلول دوم و سوم مشاهده کرد. همچنین خروجی توکنایز شده نیز خروجی‌ای قابل قبول است و این مورد نشان می‌دهد که انجام پیش‌پردازش در همین حد کفایت می‌کند.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">پیاده‌سازی N-gram</div>


<p dir='rtl' style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
یک کلاس برای ساخت و آموزش N-gram بنویسید.
<br>
با استفاده از این کلاس و مجموعه داده توکنایز شده که در بخش قبل آماده کردید، 
<span dir="ltr"> 2-gram, 4-gram, 8-gram</span>
بسازید و روی مجموعه داده خود آموزش دهید.
<br>
با استفاده از مدل‌های آموزش داده شده، متن‌های 100 توکنی تولید کنید و کیفیت متون تولیدشده را با هم مقایسه کنید و تفاوت عملکرد مدل‌ها از نظر پیوستگی و روانی متون را تحلیل کنید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- مدل‌های N-gram آموزش یافته
<br>
- متن‌های 100 توکنی تولیدشده با هر N-gram
</p>

In [1]:
import math
import random
import pickle
from collections import Counter, defaultdict
from typing import List, Optional, Tuple
import numpy as np
import pandas as pd
from tqdm import tqdm

class NGramModel:
    def __init__(self, n: int, smoothing: float = 1.0, pad_token="<s>", end_token="</s>", rng_seed: int = 42):
        self.n = n if n >= 1 else 2
        self.smoothing = float(smoothing)
        self.pad_token = pad_token
        self.end_token = end_token
        self.vocab = set()
        # numerator of P(w_i|w_i-n+1:i-1)
        self.ngram_counts = Counter()
        # denominator of P(w_i|w_i-n+1:i-1)
        self.context_counts = Counter()
        self.trained = False
        self.rng = random.Random(rng_seed)
        self.backoff = False
        self.backoff_val = 0.4
        self.interpolation = False
        self.temperature = 1.0

In [2]:
# Helper methods
def _pad_sequence(self, tokens: List[str]) -> List[str]:
        pads = [self.pad_token] * (self.n - 1)
        return pads + tokens + ([self.end_token] if self.end_token is not None else [])

def _update_counts_from_sequence(self, seq: List[str]):
    seq_len = len(seq)
    for i in range(seq_len):
        for k in range(1, self.n + 1):
            if i + k <= seq_len:
                ngram = tuple(seq[i:i + k])
                context = tuple(seq[i:i + k - 1]) if k > 1 else tuple()
                self.ngram_counts[k][ngram] += 1
                self.context_counts[k][context] += 1
                for tok in ngram:
                    self.vocab.add(tok)
            
NGramModel._pad_sequence = _pad_sequence
NGramModel._update_counts_from_sequence = _update_counts_from_sequence

In [3]:
def fit(self, token_lists: List[List[str]], verbose: bool = True):
    self.ngram_counts = {k: Counter() for k in range(1, self.n + 1)}
    self.context_counts = {k: Counter() for k in range(1, self.n + 1)}
    self.vocab = set()
    iterator = token_lists
    if verbose:
        iterator = tqdm(token_lists, desc=f"Training {self.n}-gram")
    for tokens in iterator:
        seq = list(tokens)
        padded = self._pad_sequence(seq)
        self._update_counts_from_sequence(padded)
    self.trained = True
    self.vocab_size = len(self.vocab)
    if verbose:
        print(f"Trained {self.n}-gram model. Vocab size: {self.vocab_size:,}")
            
NGramModel.fit = fit

In [4]:
import numpy as np

def _apply_temp(self, probs):
    if self.temperature != 1.0:
        probs = np.array(probs, dtype=float)
        probs = np.power(probs, 1.0 / self.temperature)
        probs = probs / probs.sum()
        probs = probs.tolist()
    return probs

def _sample_next(self, context: List[str]) -> str:
    context_full = [self.pad_token] * max(0, (self.n - 1 - len(context))) + context
    context_tuple = tuple(context_full)
    
    # interpolation
    if self.interpolation:
        lambdas = [0.4, 0.3, 0.2, 0.1]
        token_probs = {}
        for i in range(self.n, 0, -1):
            sub_context = context_tuple[-(i - 1):] if i > 1 else ()
            if i not in self.ngram_counts:
                continue
            
            # unigram
            if i == 1:
                denom = sum(self.ngram_counts[1].values()) + (self.smoothing * self.vocab_size)
                if denom == 0:
                    continue
                for ngram, cnt in self.ngram_counts[1].items():
                    token = ngram[-1]
                    prob = (cnt + self.smoothing) / denom
                    token_probs[token] = token_probs.get(token, 0) + lambdas[self.n - i] * prob
            
            else:
                context_count = self.context_counts.get(i, Counter()).get(sub_context, 0)
                denom = context_count + (self.smoothing * self.vocab_size)
                if denom == 0:
                    continue
                for ngram, cnt in self.ngram_counts[i].items():
                    if tuple(ngram[:-1]) != sub_context:
                        continue
                    token = ngram[-1]
                    prob = (cnt + self.smoothing) / denom
                    token_probs[token] = token_probs.get(token, 0) + lambdas[self.n - i] * prob
        
        if not token_probs:
            return self.pad_token
        
        tokens, probs = zip(*token_probs.items())
        
        # temperature
        probs = self._apply_temp(probs)
        
        return self.rng.choices(tokens, weights=probs, k=1)[0]
    
    candidates = []
    weights = []
    
    for ngram, cnt in self.ngram_counts[self.n].items():
        if tuple(ngram[:-1]) == context_tuple:
            candidates.append(ngram[-1])
            weights.append(cnt + self.smoothing)
    
    if len(candidates) > 0 or not self.backoff:
        denom = self.context_counts.get(self.n, Counter()).get(context_tuple, 0) + (self.smoothing * self.vocab_size)
        probs = [w / denom for w in weights]
        
        # temperautre
        probs = self._apply_temp(probs)
        
        return self.rng.choices(candidates, weights=probs, k=1)[0]
    
    # backoff
    if len(context_tuple) > 0:
        return self._sample_next(list(context_tuple[1:]))
    
    # len(context) == 0 meaning we have reached unigram level
    unigrams = []
    unigram_weights = []
    total_unigrams_count = 0
    for token in self.vocab:
        c = self.ngram_counts[1].get((token,), 0)
        if c > 0:
            unigrams.append(token)
            unigram_weights.append(c + self.smoothing)
            total_unigrams_count += c
    if not unigrams:
        return self.pad_token
    
    probs = [w / (total_unigrams_count + self.smoothing * self.vocab_size) for w in unigram_weights]
    
    # temperature
    probs = self._apply_temp(probs)
    
    return self.rng.choices(unigrams, weights=probs, k=1)[0]

NGramModel._apply_temp = _apply_temp    
NGramModel._sample_next = _sample_next


In [5]:
def generate(self, length: int = 100) -> List[str]:
        if not self.trained:
            print("Model not trained yet.")
        else:
            # start with n-1 <s>
            context = [self.pad_token] * (self.n - 1)
        
        generated = []
        for i in range(length):
            next_tok = self._sample_next(context)
            generated.append(next_tok)
            # update context
            context = (context + [next_tok])[-(self.n - 1):]
        return generated
    
NGramModel.generate = generate

In [25]:
# Save the model
def save(self, path: str):
    with open(path, "wb") as f:
        pickle.dump({
            "n": self.n,
            "smoothing": self.smoothing,
            "pad_token": self.pad_token,
            "end_token": self.end_token,
            "vocab": self.vocab,
            "ngram_counts": self.ngram_counts,
            "context_counts": self.context_counts
        }, f)

def load(cls, path: str, rng_seed = 42):
    with open(path, "rb") as f:
        data = pickle.load(f)
    model = cls(n=data["n"], smoothing=data["smoothing"], pad_token=data["pad_token"], end_token=data["end_token"], rng_seed=rng_seed)
    model.vocab = data["vocab"]
    model.ngram_counts = data["ngram_counts"]
    model.context_counts = data["context_counts"]
    model.trained = True
    model.vocab_size = len(model.vocab)
    return model

NGramModel.save = save

In [7]:
# Load train tokens
tokenized_df = pd.read_parquet('./Data/bpe_tokenized_tiny_stories_farsi.parquet')
train_tokens_list = list(tokenized_df['Persian'])
print(len(train_tokens_list))
print(type(train_tokens_list[0]), sep='\n')
print(train_tokens_list[0], sep='\n')

124411
<class 'numpy.ndarray'>
['یک' 'روز' 'یک' 'پسربچه' 'کوچولو' 'به' 'اسم' 'بن' 'بود' '.' 'بن' 'دوست'
 'داشت' 'دنیای' 'اطرافش' 'را' 'کشف' 'کند' '.' 'او' 'چیزهای' 'شگفت' 'انگیز'
 'زیادی' 'دید' '،' 'مثل' 'وازه' 'های' 'زیبایی' 'که' 'در' 'یک' 'مغازه' 'به'
 'نمایش' 'گذاشته' 'شده' 'بودند' '.' 'یک' 'روز' 'هنگامی' 'که' 'بن' 'از'
 'مغازه' 'رد' 'می' 'شد' 'یک' 'وازه' 'بسیار' 'ویژه' 'را' 'دید' '.' 'وقتی'
 'بن' 'ان' 'را' 'دید' 'مبهوت' 'شد' '!' 'او' 'گفت' ':' '«' 'واو' '،' 'این'
 'واقعا' 'یک' 'وازه' 'شگفت' 'انگیز' 'است' '!' 'ایا' 'می' 'توانم' 'ان' 'را'
 'بخرم' '؟' '»' 'فروشنده' 'لبخند' 'زد' 'و' 'گفت' ':' '«' 'بله' '،' 'البته'
 'می' 'توانی' '.' 'می' 'توانی' 'ان' 'را' 'به' 'خانه' 'ببری' 'و' 'به' 'همه'
 'دوستانت' 'نشان' 'دهی' 'که' 'چقدر' 'شگفت' 'انگیز' 'است' '!' '»' 'پس' 'بن'
 'وازه' 'را' 'به' 'خانه' 'برد' 'و' 'از' 'ان' 'بسیار' 'مفتخر' 'بود' '!'
 'او' 'دوستانش' 'را' 'صدا' 'کرد' 'و' 'وازه' 'شگفت' 'انگیز' 'را' 'به'
 'انها' 'نشان' 'داد' '.' 'همه' 'دوستانش' 'فکر' 'می' 'کردند' 'که' 'وازه'
 'زیباست' 'و' 'ن

In [9]:
for n in [8]:
    print(f'Starting training for {n}-gram...')
    model = NGramModel(n=n, smoothing=0.0)
    model.fit(train_tokens_list, verbose=True)
    model.save(f'./Trained_Ngrams/{n}gram_model.pkl')

Starting training for 8-gram...


Training 8-gram: 100%|██████████| 124411/124411 [15:03<00:00, 137.71it/s]  


Trained 8-gram model. Vocab size: 24,947


In [ ]:
# Load models
models = {}
for n in [2, 4, 8]:
    models[f'{n}-gram'] = load(NGramModel, f'./Trained_Ngrams/{n}gram_model.pkl')
    

In [53]:
for n in [2, 4, 8]:
    print(f'Text generated by the {n}-gram model:')
    generated_list = models[f'{n}-gram'].generate(length=100)
    generated_text = ' '.join(generated_list)
    print(generated_text, '\n')

Text generated by the 2-gram model:
باری روزی ، من دیگر سرسره نوک تیز داشت . می تونی کمکم کنی تا روز ، سگ خیس و عذرخواهی می خواست ان ساندویچ های زیادی داشت تمام شد که سر می زنید . انها درباره پرنده دادند و هیجان زده شد . او کمک می خواست بالای سرسره پایین افتاد . برخی صاف شد . اون خیلی زیبا توش بپرن . ولی گفت . همه دوست داشت . وقتی خواب ارام به هم دوست باشیم . ان ها در اشپزخانه کمک خواهند ان ها میشیم . او گفت : نه برای انجام می خندیدند ! می خونم برات 

Text generated by the 4-gram model:
باری روزی روزگاری ، دختری به نام لیلی زندگی می کرد . باب دوست داشت به مادر و پدرش زندگی می کرد . تام خودشو نمی شناخت . فکر کرد پرنده در انجا مشغول خواندن کتاب و خوردن میان وعده ، بادی اماده بود دوباره بازی کنه . پارک خیلی خیس شده بود که همه می توانند دوست باشند . الا گفت : بیا کیک رو با دوستاش به اشتراک بزاره ، پس یه حوله پیدا کرد و ان را بالا و پایین پرت می کرد رو جمع کرد و به دوستانش ملحق شد . انها کار را با راکت 

Text generated by the 8-gram model:
باری روزی روزگاری ، سگ کوچولویی به نام اسپات وجود د

<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش دوم:</b><br>
متن‌های تولیدی توسط هر کدام از سه مدل به صورت زیر می‌باشد (خروجی سلول بالا):
<br><br>
مدل 2-gram:
<br>
باری روزی ، من دیگر سرسره نوک تیز داشت . می تونی کمکم کنی تا روز ، سگ خیس و عذرخواهی می خواست ان ساندویچ های زیادی داشت تمام شد که سر می زنید . انها درباره پرنده دادند و هیجان زده شد . او کمک می خواست بالای سرسره پایین افتاد . برخی صاف شد . اون خیلی زیبا توش بپرن . ولی گفت . همه دوست داشت . وقتی خواب ارام به هم دوست باشیم . ان ها در اشپزخانه کمک خواهند ان ها میشیم . او گفت : نه برای انجام می خندیدند ! می خونم برات 
<br><br>
مدل 4-gram:
<br>
باری روزی روزگاری ، دختری به نام لیلی زندگی می کرد . باب دوست داشت به مادر و پدرش زندگی می کرد . تام خودشو نمی شناخت . فکر کرد پرنده در انجا مشغول خواندن کتاب و خوردن میان وعده ، بادی اماده بود دوباره بازی کنه . پارک خیلی خیس شده بود که همه می توانند دوست باشند . الا گفت : بیا کیک رو با دوستاش به اشتراک بزاره ، پس یه حوله پیدا کرد و ان را بالا و پایین پرت می کرد رو جمع کرد و به دوستانش ملحق شد . انها کار را با راکت 
<br><br>
مدل 8-gram:
<br>
باری روزی روزگاری ، سگ کوچولویی به نام اسپات وجود داشت . اسپات دوست داشت روزهایش را با بازی با توپش سپری کند . او می دوید و می پرید و ریسمان را در دهانش می گرفت . مکس و ریسمان بهترین دوستان بودند . یک روز ، تام و تیم در پارک بازی می کردند . ان ها دوست داشتند بدوند ، بپرند و به دنبال هم بدوند . انها همچنین دوست داشتند به اسمان نگاه کنند و اشکال مختلف ابرها را ببینند . گاهی اوقات حیوانات ، یا چهره ها ، یا قلعه ها می دیدند . لیلی 
<br><br>
همانطور که به وضوح دیده می‌شود، متن تولیدی توسط مدل 4-gram بهتر از مدل 2-gram و مدل 8-gram به مراتب بهتر از دو مدل دیگر می‌باشد تا جایی که می‌توان از آن یک داستان را تا حدی برداشت کرد و از پیوستگی مفهومی قابل توجهی برخوردار است؛ جمله‌ها معنی دارند، پیوسته اند و انسجام مفهوم تاحد خیلی خوبی در طول متن حفظ شده است. همچنین مدل 4-gram نیز نسبت به مدل 2-gram پیشرفت چشمگیری داشته و انسجام خیلی بهتری در تولید متن مورد نظر از خود نشان داده است در حالی که مدل 2-gram عملکرد ضعیفی داشته و نمی‌توان هیچ گونه مفهوم و یا معنی‌ای از آن استخراج کرد.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">معیار Perplexity</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابتدا از مجموعه داده 
<a href="https://huggingface.co/datasets/taesiri/TinyStories-Farsi" target="_blank" rel="noopener noreferrer" style="color: #9EEAD2; text-decoration: underline;">
    TinyStories-Farsi
</a>
دادگان validation فارسی را جدا کنید.
<br>
سپس معیار Perplexity را برای هر کدام از N-gram های خود، روی این مجموعه حساب کنید.
<br>
تحلیل خود را از این نتیجه بگویید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- مجموعه دادگان فارسی validation
<br>
- نتیجه معیار Perplexity در هر N-gram
</p>

In [24]:

def _prob_ngram(self, ngram: Tuple[str, ...], eps='eps', first_backoff=True, return_eps_str=True) -> float:
    if not self.trained:
        print("Model is not trained yet.")
    assert 1 <= len(ngram) <= self.n

    k = len(ngram)
    if k == 1:
        count = self.ngram_counts[1].get(ngram, 0)
        denom = sum(self.ngram_counts[1].values()) + (self.smoothing * self.vocab_size)
        if denom == 0:
            return eps if return_eps_str else 1e-8
        return (count + self.smoothing) / denom

    context = tuple(ngram[:-1])
    token = ngram[-1]
    context_count = self.context_counts[k].get(context, 0)
    ngram_count = self.ngram_counts[k].get(ngram, 0)

    # interpolation
    if self.interpolation:
        lambdas = [0.4, 0.3, 0.2, 0.1]
        prob = 0.0
        for i in range(k, 0, -1):
            sub_ngram = ngram[-i:]
            count = self.ngram_counts[i].get(sub_ngram, 0)
            denom = self.context_counts[i].get(sub_ngram[:-1], 0) + self.smoothing * self.vocab_size
            p = (count + self.smoothing) / denom if denom > 0 else 0.0
            prob += lambdas[self.n - i] * p
        return prob if prob > 0 else eps

    if context_count > 0 or self.smoothing > 0.0:
        numerator = ngram_count + self.smoothing
        denominator = context_count + self.smoothing * self.vocab_size
        return numerator / denominator

    elif not self.backoff:
        return eps if return_eps_str else 1e-8

    # backoff
    else:
        return self.backoff_val * self._prob_ngram(tuple(ngram[1:]), eps)
    
NGramModel._prob_ngram = _prob_ngram

In [23]:
def score(self, token_list: List[str], add_start_end: bool = True, eps='eps', return_eps_str=True) -> float:
        if add_start_end:
            seq = self._pad_sequence(token_list)
        else:
            seq = list(token_list)
        
        pred_count = 0
        total_logp = 0.0
        seq_len = len(seq)
        for i in range(seq_len - self.n + 1):
            ngram = tuple(seq[i:i + self.n])
            prob = self._prob_ngram(ngram, return_eps_str=return_eps_str)
            if prob != eps and prob <= 0:
                prob = 1e-20
            if prob != eps:
                pred_count += 1
                total_logp += math.log(prob)
        return total_logp, pred_count
    
NGramModel.score = score

As we have seen in the course's slides, the classical equation for Perplixty is defined as: 
$$
\text{PP}(W) = \left( \prod_{i=1}^{N} \frac{1}{P(w_i \mid w_{i-1})} \right)^{\frac{1}{N}}
$$
For simpler and more robust computation, let's first take Log from both sides:
$$
\log \text{PP}(W) = \frac{1}{N} \sum_{i=1}^{N} -\log P(w_i \mid w_{i-1})
$$
And now to turn it back to Perplexity's value:
$$
\text{PP}(W) = \exp\left( -\frac{1}{N} \sum_{i=1}^{N} \log P(w_i \mid w_{i-1}) \right)
$$
And this is the form that we will compute in the method below.

In [22]:
def perplexity(self, token_lists: List[List[str]], add_start_end: bool = True, return_eps_str=True) -> float:
        total_logp = 0.0
        total_tokens = 0
        total_no_context = 0
        for tokens in token_lists:
            tokens = list(tokens)
            if add_start_end:
                seq = self._pad_sequence(tokens)
            else:
                seq = tokens
            
            # number of predicted tokens is len(seq) - (n-1) as first contexts don't predict
            dif = max(0, len(seq) - (self.n - 1))
            if dif == 0:
                continue    
            
            logp, pred_count = self.score(tokens, add_start_end=add_start_end, return_eps_str=return_eps_str)
            total_tokens += pred_count
            total_logp += logp
            total_no_context += dif - pred_count if dif != 0 else 0
        
        if total_tokens == 0:
            return float('inf')
        
        avg_neg_logp = - total_logp / total_tokens
        return math.exp(avg_neg_logp), sum([len(tokens) - self.n + 1 for tokens in token_lists]), total_no_context

NGramModel.perplexity = perplexity

In [ ]:
import pandas as pd

# get the validation data
splits = {'train': 'InProgress-TinyStoriesV2-GPT4-train-Translated-To-Farsi-w-Claude-2.0.csv', 'validation': 'TinyStoriesV2-GPT4-valid-Translated-To-Farsi-w-Claude-2.0.csv'}
df = pd.read_csv("hf://datasets/taesiri/TinyStories-Farsi/" + splits["validation"])
print(df.columns)
df.to_parquet('./Data/Tiny_Stories_Farsi_Validation.parquet')

Index(['English', 'Persian'], dtype='object')


In [19]:
val_df = pd.read_parquet('./Data/Tiny_Stories_Farsi_Validation.parquet')
val_df.drop(columns=['English'], inplace=True, axis=1)
print(val_df.shape)
print(val_df.columns)

(27630, 1)
Index(['Persian'], dtype='object')


Now, to preprocess and tokenize the validation set:

In [ ]:
from parsivar import Normalizer
import re

def preprocess_text(text, normalizer):
    # Normalize text
    text = normalizer.normalize(text)
    # Replace half space with space for clarity
    text = text.replace('\u200c', ' ')
    # Remove Latin letters
    text = re.sub(r'[A-Za-z0-9]', '', text)
    # Remove stray symbols or quotes
    text = re.sub(r'[“”"\'`]', '', text)
    # Collapse multiple spaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

def preprocess_data(df, preprocess_function, normalizer, col_name='Persian'):
    cleaned_df = df.copy()
    cleaned_df[col_name] = cleaned_df[col_name].apply(lambda x: preprocess_function(x, normalizer))      
    return cleaned_df


normalizer = Normalizer(statistical_space_correction=True)
cleaned_val_df = preprocess_data(val_df, preprocess_text, normalizer)

In [21]:
from tokenizers import Tokenizer

def bpe_tokenize_data(cleaned_df, tokenizer, col_name='Persian'):
    tokenized_df = cleaned_df.copy()
    tokenized_df[col_name] = cleaned_df[col_name].apply(lambda x: tokenizer.encode(x).tokens)
    return tokenized_df

bpe_tokenizer = Tokenizer.from_file('./Tokenizer/tiny_stories_farsi_bpe.json')
tokenized_val_df = bpe_tokenize_data(cleaned_val_df, bpe_tokenizer)
print(tokenized_val_df.shape, tokenized_val_df.columns)
print(type(tokenized_val_df.iloc[0]['Persian']))
print(tokenized_val_df.iloc[0]['Persian'])

(27630, 1) Index(['Persian'], dtype='object')
<class 'list'>
['ترجمه', 'به', 'فارسی', ':', 'مجبور', 'نیستی', 'از', 'سگ', 'بلند', 'صدا', 'ترسیده', 'باشی', '،', 'من', 'ازت', 'محافظت', 'میکنم', '.', 'خرگوش', 'خاکی', 'با', 'دخترک', 'احساس', 'امنیت', 'زیادی', 'می', 'کرد', '.', 'او', 'خیلی', 'مهربان', 'بود', 'و', 'خرگوش', 'خاکی', 'به', 'زودی', 'به', 'او', 'اعتماد', 'پیدا', 'کرد', '.', 'او', 'به', 'دخترک', 'تکیه', 'داد', 'و', 'دخترک', 'از', 'او', 'محافظت', 'کرد', '.', 'خرگوش', 'خاکی', 'بهترین', 'دوستش', 'را', 'پیدا', 'کرده', 'بود', '.']


In [22]:
tokenized_val_df.to_parquet('./Data/bpe_tokenized_tiny_stories_farsi_Valid.parquet')

Now, we can compute the perplexity of the three models:

In [56]:
# Load models
models = {}
for n in [2, 4, 8]:
    models[f'{n}-gram'] = load(NGramModel, f'./Trained_Ngrams/{n}gram_model.pkl')

In [10]:
train_tokenized_df = pd.read_parquet('./Data/bpe_tokenized_tiny_stories_farsi.parquet')
train_tokens_list = list(train_tokenized_df['Persian'])
tokenized_val_df = pd.read_parquet('./Data/bpe_tokenized_tiny_stories_farsi_Valid.parquet')
valid_tokens_list = list(tokenized_val_df['Persian'])

In order to resolve the challenge of seeing a token that does not appear in the training set without using smoothing, backoff, or interploation, let's replace any token in the valid vocabulary that is not present in the train vocab with a singletone token of the train vocab:

In [56]:
from collections import Counter
import random

def replace_unseen_with_singleton(train_token_lists, val_token_lists, randomize=False):
    freq = Counter(token for sent in train_token_lists for token in sent)    
    singletons = [tok for tok, c in freq.items() if c == 1]
    
    if not singletons:
        print("No singletones in the train vocab")
        return
    
    fixed_singleton = random.choice(singletons)
    vocab = set(freq.keys())
    train_processed = train_token_lists    
    val_processed = []
    for sent in val_token_lists:
        new_sent = []
        for tok in sent:
            if tok in vocab:
                new_sent.append(tok)
            else:
                if randomize:
                    new_sent.append(random.choice(singletons))
                else:
                    new_sent.append(fixed_singleton)
        val_processed.append(new_sent)
    
    return val_processed

valid_tokens_list = replace_unseen_with_singleton(train_tokens_list, valid_tokens_list, randomize=True)
print(len(valid_tokens_list))

27630


In [94]:
for n in [2, 4, 8]:
    pp, total_seqs, total_unseen_seqs = models[f'{n}-gram'].perplexity(valid_tokens_list)
    print(f"{n}-gram perplexity on validation set: {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

2-gram perplexity on validation set: 158.613, percentage of unseen seqs for prediction: 0.00
4-gram perplexity on validation set: 88470.353, percentage of unseen seqs for prediction: 14.42
8-gram perplexity on validation set: 2684393.406, percentage of unseen seqs for prediction: 79.40


In order to be compatible with the models that use smoothing, let's set the return_eps_str parameter off once and get the perpelxity in order to be used for the upcomings comparison:

In [ ]:
model = load(NGramModel, f'./Trained_Ngrams/4gram_model.pkl')
# model = models[4-gram]
pp, total_seqs, total_unseen_seqs = model.perplexity(valid_tokens_list, return_eps_str=False)
print(f"4-gram perplexity on validation set (returns an eps value on no_context): {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

4-gram perplexity on validation set (returns an eps value on no_context): 238711.274, percentage of unseen seqs for prediction: 0.00


<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش سوم:</b><br>
همانطور که دیده می‌شود، مقدار perplexity برای مدل 2-gram نسبت به دو مدل دیگر به مراتب کمتر است. علت این پدیده این است که همان‌طور که ملاحظه می‌شود، در مدل 4-gram و مخصوصا مدل 8-gram، قسمت زیادی از دادۀ validation در مجموعۀ آموزش وجود نداشته است؛ به عبارت دیگر، در این مدل، کانتکست n-gram که برای مدل 8-grm شامل 7 و برای مدل 4-gram شامل 3 توکن است، در تعداد قابل توجهی از اوقات در مجموعه دادۀ آموزش وجود نداشته است و همان‌طور که در پیاده‌سازی دیده می‌شود، در این مواقع تصمیم گرفته شده است که آن نمونه در ارزیابی در نظر گرفته نشود، و به همین دلیل معیار perplexity با تعداد بسیار کمتری نسبت به تعداد دادۀ واقعی اندازه‌گیری شده و مقدار آن در مقایسه با مدل 2-gram که تمامی ترکیبات در دادۀ آموزش وجود داشته است، بسیار بزرگتر و ناپایدارتر می‌باشد.
به این دلیل (حساسیت روی طول پنجرۀ محتوا) و دلایل دیگر، معیار perplexity لزوما بیانگر قدرت مدل نمی‌باشد و همان‌طور که دیده می‌شود، قدرت مدل 8-gram در مقایسه با مدل 2-gram در تولید متن به مراتب بیشتر بوده است در حالی که این مورد برای مقدار perplexity صادق نیست. به طور خلاصه، به علت اینکه در تعداد قابل توجهی از نمونه‌های دادۀ validation، کنتکست مورد نظر در دادۀ آموزش وجود نداشته است، احتمال بازگردانده شدده باعث افزایش چشمگیر مقدار perplexity می‌شود که برای حل این مشکل باید از یکی از روش‌های smoothing بهره جست.
</p>


<p dir='rtl' style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در این قسمت چهار جمله در اختیار شما قرار گرفته است. استدلال کنید کدام جمله‌ها محتمل‌ترند که توسط یک <span dir="ltr">4-gram</span> که روی داده‌های مشابه آموزش دیده‌است، تولید شده باشند.
<br>
جمله‌ی اول = آنها دوست داشتند در ماسه بازی کنند و جزر و مد آب را تماشا کنند
<br>
جمله‌ی دوم = جیل و تام به همراه مامان و بابا به ساحل رفتند
<br>
جمله‌ی سوم = تام بطری‌ نوشابه‌ را تا حد ممکن بالا انداخت و به سمت او فریاد زد
<br>
جمله‌ی چهارم = باری خیلی دوست داشت بیرون از منزل نقاشی کند و با پدربزرگ منظره تماشا کند
</p>

<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش چهارم:</b><br>
جملۀ دوم محتمل‌تر است. مدل 4-gram برای مدل کردن و فهم ارتباط توالی‌هایی به طول 4 آموزش می‌بیند و و می‌تواند ارتباط توکن‌ها تا چهار توکن بعدی را درک کند و در نتیجه، متون خروجی آن معمولا به صورت تک‌جمله‌هایی کوتاه تشکیل شده از کلمات رایج می‌باشد. حال آن که این مشخصات در گزینۀ دوم دیده می‌شود. به عبارت دیگر، در گزینۀ دوم ترکیباتی همچون "به ساحل رفتند" و یا "به همراه مامان" که متشکل از سه الی چهار توکن هستند بسیار شبیه به  توالی‌هایی است که یک مدل 4-gram این موارد گزینۀ دوم را گزینۀ محتمل‌تر می‌کند. در مجموعۀ داده و توالی‌های چهارتایی خود می‌بیند و 
<br>
در گزینۀ اول، تشکیل شدن از دو جمله و وجود ارتباط بین توکن و آخر (فاعل و فعل جملۀ دوم) احتمال اینکه این گزینه حاصل خروجی یک مدل 4-gram باشد را کم می‌کند، چرا که با توجه به طول توالی‌هایی که این مدل بر روی آن آموزش می‌بیند، توانایی در مدل کردن ارتباط بین توکن اول و آخر این گزینه از این مدل انتظار نمی‌رود.
<br>
گزینۀ سوم نیز همانند گزینۀ اول شامل وجود ارتباط بین توکن اول و فعل آخر جمله است که احتمال تولید شدن توسط یک مدل 4-gram تا حداکثر توالی به طول 4 را می‌تواند ببیند، کم می‌کند. 
<br>
در جملۀ چهارم نیز همان موردی که برای جملۀ اول و سوم گفته شد، وجود دارد که احتمال تولید شدن توسط مدل 4-gram را کاهش می‌دهد. به طور کلی نیز، در یک مدل 4-gram، همانطور که گفته شد، معمولا جملاتی کوتاه‌تر با ترکیباتی رایج‌تر دیده می‌شود که وجود ترکیبات کمیاب‌تر در گزینه‌های اول، سوم و چهارم همانند "جزر و مد" و یا افعال و جملات مرکب (مخصوصا در گزینۀ چهارم که یک مرکب است که هر جمله ساختار درستی دارد) احتمال تولید شدن توسط مدل 4-gram را کاهش می‌دهد.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">روش‌های هموارسازی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
هر کدام از الگوریتم‌های هموارسازی (Smoothing) که در درس خوانده‌اید (Laplace, Interpolation, Backoff) را پیاده‌سازی کنید.
<br>
یک مدل
<span dir="ltr">4-gram</span>
بسازید و با مجموعه داده توکنایزشده فارسی که در بخش اول این سوال آماده کردید، آموزش دهید.
(می‌توانید از مدل 
<span dir="ltr">4-gram</span>
آموزش‌یافته قبلی خود استفاده کنید.)
<br>
در الگوریتم Interpolation مقادیر λ را 0.4, 0.3, 0.2, 0.1 در نظر بگیرید. به این صورت:
P​(w∣h3​,h2​,h1​)=0.4P​(w∣h3​,h2​,h1​)+0.3P​(w∣h2​,h1​)+0.2P​(w∣h1​)+0.1P​(w)
<br>
در الگوریتم Backoff مقدار λ را 0.4 در نظر بگیرید.
<br>
سپس با استفاده از هر یک از این روش‌ها، متن‌هایی به طول 100 توکن تولید کنید.
<br>
متن‌های تولیدشده را با یکدیگر مقایسه کنید و تفاوت آن‌ها را از نظر روانی و تنوع کلمات بررسی کنید.
<br>
در نهایت، با استفاده از احتمالات محاسبه‌شده در هر یک از این روش‌ها، معیار Perplexity را روی مجموعه دادگان فارسی validation - که در بخش سوم این سوال آماده کردید - حساب کنید.
<br>
مقادیر Perplexity برای هر کدام از روش‌های هموارسازی و مدل غیرهموار (Unsmoothed) را با یکدیگر مقایسه کنید و تحلیل خود را از این نتایج بگویید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- متن‌های تولیدشده با هر یک از روش‌های هموارسازی ذکر شده
<br>
- معیار Perplexity برای هر یک از روش‌های هموارسازی ذکر شده
</p>

As seen in the above implemented class, the infrastructure for backoff and smoothing is already implemented and we just need to assign their values:

In [ ]:
model_4gram = load(NGramModel, './Trained_Ngrams/4gram_model.pkl')
# model_4gram = models[4-gram]

#### **Laplacian (Add-1)**

In [120]:
model_4gram.smoothing = 1.0
model_4gram.backoff = False
model_4gram.interpolation = False

In [121]:
generated_list = model_4gram.generate(length=100)
generated_text = ' '.join(generated_list)
print(generated_text, '\n')

باری روزی روزگاری ، گربه بزرگی موش کوچکی را در خارج از خانه در گل و لای بیرون بکشند . ککی ازاد شد و انها با هم کار کردن و چوب ها ساخته بود . این قهرمان بسیار راستگو بود . یک روز ، لیلی با مادرش به فروشگاه رفتند . انها به راهشان ادامه دادند . تیم خوشحال بود و شروع به تماشای حیوانات مختلف بنشیند . یک روز ، زیتون یک دانه کوچک روی زمین دید . او گفت : شما اژیر اسباب بازی من گم شده ام و بشکه ام در حال نشت کردن است . لاک 



In [59]:
pp, total_seqs, total_unseen_seqs = model_4gram.perplexity(valid_tokens_list)
print(f"4-gram perplexity with smoothing: {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

4-gram perplexity with smoothing: 2035.881, percentage of unseen seqs for prediction: 0.00


#### **Backoff**

In [133]:
model_4gram.smoothing = 0.0
model_4gram.interpolation = False
model_4gram.backoff = True

In [103]:
generated_list = model_4gram.generate(length=100)
generated_text = ' '.join(generated_list)
print(generated_text, '\n')

یک روزی یک سگ کوچولو به نام لیلی زندگی می کرد . اسپات دوست داشت کل روز بخوابد . یک روز ، دختربچه ای به نام تیم او را خیلی دوست داشت بدوه . یه روز ، بابا خانواده را به رستوران ببرد . مادر گفت : متشکرم ، تام می خواست یک ایگلو درست کند . لیلی و بن از کارکنان عذرخواهی می کنند . انها بسیار خوشحال بودند که یک پرنده حرف می زد . یک روز ، توپ وارد جعبه شد . او پرواز کرد . تیم دوست داشت با کمانش بازی کند . او رکس را 



In [139]:
pp, total_seqs, total_unseen_seqs = model_4gram.perplexity(valid_tokens_list)
print(f"4-gram perplexity with backoff (discount value = 0.4): {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

4-gram perplexity with backoff (discount value = 0.4): 186442.546, percentage of unseen seqs for prediction: 0.00


In [66]:
discount_vals = [0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0]
for discount_val in discount_vals:
    model_4gram.backoff_val = discount_val
    pp, total_seqs, total_unseen_seqs = model_4gram.perplexity(valid_tokens_list)
    print(f"4-gram perplexity with backoff (discount value = {discount_val}): {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

4-gram perplexity with backoff (discount value = 0.6): 181436.267, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 0.8): 175232.251, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 1.0): 170566.511, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 1.2): 166846.693, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 1.4): 163764.994, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 1.6): 161141.552, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 1.8): 158862.418, percentage of unseen seqs for prediction: 0.00
4-gram perplexity with backoff (discount value = 2.0): 156850.991, percentage of unseen seqs for prediction: 0.00


This flow suggests that most of the data samples in the validation set enters backoff and also, mostly backs off all the way to the unigram level.

#### **Interpolation**

We now also add the support for interpolation in the main class (code above).

In [129]:
model_4gram.interpolation = True
model_4gram.backoff = False
model_4gram.smoothing = 0.0

In [112]:
generated_list = model_4gram.generate(length=100)
generated_text = ' '.join(generated_list)
print(generated_text, '\n')

یک روز یک گربه سرزنده به نام انا و مادرش کرم در حال کشیدن هویج است . خوبی ، لیلی . دامپزشک مودبانه بازی را پس را لمس کنی و می شد . دیگر نمی خواستند بازی کنند . انها با رنگ های زیبا و میوه کرد ! عطسه ! باب عطسه کرد شدند ، ایمی گفت . به نظر غمگین می امد . خوشحالم . می گوید : فرض کنیم در یک درخت بزرگ نزدیک حفره عمیق افتاد یادداشتی در داخل ان ، تام خوشحال بود که از بپرد و ! ایمی گفت : البته ! فقط یه تزریق 



In [130]:
pp, total_seqs, total_unseen_seqs = model_4gram.perplexity(valid_tokens_list)
print(f"4-gram perplexity with interpolation: {pp:.3f}, percentage of unseen seqs for prediction: {(100 * total_unseen_seqs / total_seqs):.2f}")

4-gram perplexity with interpolation: 33.273, percentage of unseen seqs for prediction: 0.00


#### **Answer**

<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش پنجم:</b><br>
متن تولید شده در مدل اولیه بدون smoothing و برای حالات smoothing به همراه مقدار perplexity در هر سه حالت را با یکدیگر مقایسه می‌کنیم:
<br><br>
No Smoothing:
<br>
متن تولید شده:
<br>
باری روزی روزگاری ، دختری به نام لیلی زندگی می کرد . باب دوست داشت به مادر و پدرش زندگی می کرد . تام خودشو نمی شناخت . فکر کرد پرنده در انجا مشغول خواندن کتاب و خوردن میان وعده ، بادی اماده بود دوباره بازی کنه . پارک خیلی خیس شده بود که همه می توانند دوست باشند . الا گفت : بیا کیک رو با دوستاش به اشتراک بزاره ، پس یه حوله پیدا کرد و ان را بالا و پایین پرت می کرد رو جمع کرد و به دوستانش ملحق شد . انها کار را با راکت 
<br>
Perplexity:238711.274
<br><br>
Laplacian:
<br>
متن تولید شده:
<br>
باری روزی روزگاری ، گربه بزرگی موش کوچکی را در خارج از خانه در گل و لای بیرون بکشند . ککی ازاد شد و انها با هم کار کردن و چوب ها ساخته بود . این قهرمان بسیار راستگو بود . یک روز ، لیلی با مادرش به فروشگاه رفتند . انها به راهشان ادامه دادند . تیم خوشحال بود و شروع به تماشای حیوانات مختلف بنشیند . یک روز ، زیتون یک دانه کوچک روی زمین دید . او گفت : شما اژیر اسباب بازی من گم شده ام و بشکه ام در حال نشت کردن است . لاک 
<br>
Perplexity: 2035.881
<br>
--> توضیحات: همانطور که دیده می‌شود، متن تولید شده در روش Laplac نسبت به حالتی که از روش Smoothing استفاده نکرده بودیم، پیشرفت چندانی نداشته است و حتی در برخی موارد، روانی و پیوستگی متن، افت نیز داشته است. این مورد به این دلیل است که در حالتی که از روش Laplac استفاده می‌کنیم، به علت اضافه کردن مقدار قابل توجهی به مخرج احتمال، احتمال انتخاب کلمات مورد نظر و قابل قبول به مقدار قابل توجهی کاهش می‌یابد و همین مورد باعث می‌شود در تولید متن، روانی و پیوستگی تا حدی از دست برود. از طرف دیگر، همان طور که دیده می‌شود، مقدار perplexity نسبت به مقدار آن در حالت بدون smoothing کاهش قابل توجهی داشته است، که این به دلیل این است که با استفاده از روش Laplac، دیگر مخرج کسر در محاسبۀ احتمال صفر نشده و در نتیجه مقدار احتمال خیلی کمی بازگردانده نمی‌شود و یا از در نظر گرفتن ان نمونه جلوگیری می‌شود و در نتیجه به علت وجود علامت منفی، مقدار توان در عبارت محاسبۀ مقدار perplexity کاهش یافته و در نتیجه مقدار نهایی perplexity کاهش قابل توجهی می‌یابد.
<br><br>
Backoff:
<br>
متن تولید شده:
<br>
یک روزی یک سگ کوچولو به نام لیلی زندگی می کرد . اسپات دوست داشت کل روز بخوابد . یک روز ، دختربچه ای به نام تیم او را خیلی دوست داشت بدوه . یه روز ، بابا خانواده را به رستوران ببرد . مادر گفت : متشکرم ، تام می خواست یک ایگلو درست کند . لیلی و بن از کارکنان عذرخواهی می کنند . انها بسیار خوشحال بودند که یک پرنده حرف می زد . یک روز ، توپ وارد جعبه شد . او پرواز کرد . تیم دوست داشت با کمانش بازی کند . او رکس را 
<br>
Perplexity: 190555.084
<br>
--> توضیحات: ملاحظه می‌شود که متن تولیدی در مقایسه با متن تولید شده در روش Laplac، متنی منسجم‌تر و تا حدی روان‌تر است و ساختاری شبیه به متن تولید شده در حالت عدم استفاده از smoothing دارد. از این مورد نتیجه می‌شود که در تولید متن، روش backoff روش نسبتا بهتری از روش Laplac می‌تواند باشد، که این مورد در اسلایدهای درس هم اشاره شده بود. در واقع در روش backoff، به علت اینکه در صورت موجود نبودن کانتکست مورد نظر در سطح کنونی به سطح پایینی می‌رویم و از همان توزیع موجود در سطح پایینی استفاده می‌شود، آرایش توزیع تا حدی حفظ می‌شود و مشکلی که در روش Laplac ذکر شد، کمتر خود را نشان می‌دهد.
<br>
در مورد perplexity همان‌طور که دیده می‌شود، مقدار آن نسبت به روش no smoothing کاهش یافته است اما نسبت به روش Laplac بیشتر است. علت این پدیده می‌تواند این باشد که به دلیل موجود نبودن بخش عمده‌ای از کانتکست در سطح چهارم، در بسیاری از مواقع به سطح‌های پایینی ارجاع داده می‌شود که در اثر آن، ضریب backoff چند بار در خود ضرب می‌شود و از آن‌جایی که احتمال حاصل شده در سطح نهایی نیز عدد لزوما بالایی نیست، عدد نهایی احتمال مقدار خیلی کمی را دارا می‌شود و در نتیجه بدین ترتیب، مقدار نهایی توان در محاسبۀ perplexity و مقدار خود این معیار افزایش می‌یابد.
<br>
همانطور که دیده می‌شود که با افزایش مقدار ضریب backoff، مقدار perplexity کاهش قابل توجهی می‌یابد که این خود می‌تواند گواهی بر ادعای عنوان شده و موجود نبودن کانتکست تعداد زیادی از نمونه‌ها در سطح چهارم و در نتیجه آن ضرب شدن متعدد ضریب backoff در خود باشد.
<br><br>
Interpolation:
<br>
متن تولید شده:
<br>
یک روز یک گربه سرزنده به نام انا و مادرش کرم در حال کشیدن هویج است . خوبی ، لیلی . دامپزشک مودبانه بازی را پس را لمس کنی و می شد . دیگر نمی خواستند بازی کنند . انها با رنگ های زیبا و میوه کرد ! عطسه ! باب عطسه کرد شدند ، ایمی گفت . به نظر غمگین می امد . خوشحالم . می گوید : فرض کنیم در یک درخت بزرگ نزدیک حفره عمیق افتاد یادداشتی در داخل ان ، تام خوشحال بود که از بپرد و ! ایمی گفت : البته ! فقط یه تزریق 
<br>
Perplexity: 33.273
<br>
--> توضیحات: می‌توان گفت که متن تولید شده در روش interpolation، نسبت به روش backoff و روش No smoothing، از انسجام و روانی کمتری برخوردار است. یکی از دلایل این مورد این می‌تواند باشد که در روش interpolation، توزیع احتمال توکن‌ها در سطوح پایین تر نیز در محاسبۀ احتمال توکن در سطح چهارم مدل نیز تاثیرگذار اند و این مورد، کار مدل را در حفظ ارتباط بین 4 توکن متوالی، موضوعی که از مدل 4-gram نتظار می‌رود، سخت‌تر می‌کند. به عبارت دیگر، تاثیر گذاشتن احتمال توکن‌ها در سطوح 1، 2 و 3-gram باعث می‌شود مدل نسبت به حالتی که تنها احتمال در سطح چهارم را مورد بررسی قرار می‌داد، سخت‌تر بتواند ارتباط بین چهار توکن متوالی را حفظ کند و ارتباط توکن‌ها در سطوح دیگر نیز در انتخاب توکن بعدی دخیل می‌شود که این سبب کاهش نسبی انسجام و روانی متن می‌شود که در متن تولیدی دیده می‌شود.
<br>
در مورد perplexity اما، مطابق انتظار، استفاده از روش perplexity بسیار بهتر از دو روش دیگر عمل کرده است و perplexity مدل را از مقدار 238711.274 به مقدار 33.273 رسانده است که این  به علت این است که در محاسبۀ احتمال هر توکن، احتمال آن در هر چهار سطح دخیل هستند و در نتیجه مقادیر نسبتا بالایی از احتمال تولید می‌شود که این در نهایت سبب کاهش مقدار perplexity می‌شود.
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">پیاده‌سازی Temperature</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در مورد تاثیر Temperature در مدل‌های زبانی توضیح‌دهید.
<br>
به هنگام Sampling از N-gram، برای آن Temperature تعریف کنید و پیاده‌سازی‌های لازم را انجام دهید.
<br>
با قرار دادن کمترین Temperature و با بیشترین Temperature، سه‌بار خروجی‌هایی به طول ۲۰ توکن تولید کنید. (برای هر حالت سه‌بار) و سپس تاثیر Temperature در خروجی‌ها را تحلیل کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- متن‌های تولید‌شده با Temperature بالا و پایین (برای هریک ۳ عدد متن تولید شده)
</p>

In [12]:
model_4gram.smoothing = 0.0
model_4gram.backoff = False
model_4gram.interpolation = False

In [162]:
temp_vals = [0.01, 0.5, 1.0]
for temp in temp_vals:
    print(f'At temperature {temp}:')
    model_4gram.temperature = temp
    for i in range(3):
        print(f'-> Text number {i + 1}:')
        generated_list = model_4gram.generate(length=20)
        generated_text = ' '.join(generated_list)
        print(generated_text)
    print('')

At temperature 0.01:
-> Text number 1:
یک روز ، یک گربه بزرگ و یک گربه کوچک به نام تیم بود . تیم دوست داشت با دوستانش
-> Text number 2:
یک روز ، یک گربه بزرگ و یک گربه کوچک به نام تیم بود . تیم دوست داشت با دوستانش
-> Text number 3:
یک روز ، یک گربه بزرگ و یک گربه کوچک به نام تیم بود . تیم دوست داشت با دوستانش

At temperature 0.5:
-> Text number 1:
یک روزی ، یک پسر بچه ای به نام تیم بود . تیم دوست داشت در پارک بازی می کرد
-> Text number 2:
یک روزی ، در یک جنگل بزرگ ، یک پرنده کوچک دید . پرنده غمگین بود . او از مادرش
-> Text number 3:
یک روز ، یک پرنده کوچک به نام تیم در حیاط خانه اش پیدا کرد . او یک گاز بزرگ

At temperature 1.0:
-> Text number 1:
با عرض پوزش ، متن انگلیسی را به فارسی ترجمه می کنم . قهرمان اسباب بازی را از اب گرم
-> Text number 2:
روزی روزگاری ، یک پسر بچه ای بود به اسم سو بود . سو دوست داشت اشپزی کند . او
-> Text number 3:
یک روز یکی بود ، یکی نبود ، یه دختری خوش شانس هستی ! » پرنده پایین نگاه کرد ،



In [13]:
temp = 3.0
model_4gram.temperature = temp
print(f'At temperature {temp}:')
for i in range(3):
    print(f'-> Text number {i + 1}:')
    generated_list = model_4gram.generate(length=20)
    generated_text = ' '.join(generated_list)
    print(generated_text)

At temperature 3.0:
-> Text number 1:
دیزی و مامان یکدیگر را بغل کنیم و ببوسیم . » لیزا شگفت زده و ناراحت . مامان ! این
-> Text number 2:
رندی یک گربه چاق رو دید و خندید وقتی ان را بخوری ، باشه ؟ ماکس پاسخ داد : اسباب
-> Text number 3:
دیوید و سوزی برادر و خواهری بودند که بسیار هیجان زده بود و همیشه اطراف را نگاه کند . ماه


<p dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی زیربخش ششم:</b><br>
همان‌طور که دیده می‌شود، با کاهش مقدار perplexity، متن تولید شده در هر سه بار به یکدیگر نزدیک و نزدیک‌تر می‌شود و رفتار مدل deterministicتر می‌شود تا جایی که در هنگامی که temperature برابر با 0.01 است، هر سه متن تولید شده عینا یکسان هستند. با افزایش temperature به مقدار 0.5، متون تولیدی از حالت یکسان خارج می‌شوند و تنها چند توکن اولیۀ آن‌ها یکسان می‌باشند. با افزایش بیشتر مقدار temperature به 1 و سپس 3 نیز مشاهده می‌شود که رفتار مدل non-deterministic شده و توکن‌های متن‌های تولید شده در حالتی که مقدار temperature برابر با 3 است، کمترین شباهتی با یکدیگر ندارند که نشان از رندوم و non-deterministic بودن رفتار مدل است.
</p>


# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>